lighthouse-core/gather/computed/critical-requests.js

```diff
@@ -0,0 +1,112 @@
+/**
+ * @license Copyright 2016 Google Inc. All Rights Reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
+ */
+
+'use strict';
+
+const ComputedArtifact = require('./computed-artifact');
+const WebInspector = require('../../lib/web-inspector');
+
+class CriticalRequests extends ComputedArtifact {
+
+  get name() {
+    return 'CriticalRequests';
+  }
+
+  /**
+   * For now, we use network priorities as a proxy for "render-blocking"/critical-ness.
+   * It's imperfect, but there is not a higher-fidelity signal available yet.
+   * @see https://docs.google.com/document/d/1bCDuq9H1ih9iNjgzyAL0gpwNFiEP4TZS-YLRp_RuMlc
+   * @param  {any} request
+   */
+  static isCritical(request) {
+    const resourceTypeCategory = request._resourceType && request._resourceType._category;
+
+    // XHRs are fetched at High priority, but we exclude them, as they are unlikely to be critical
+    // Images are also non-critical.
+    // Treat any images missed by category, primarily favicons, as non-critical resources
+    const nonCriticalResourceTypes = [
+      WebInspector.resourceTypes.Image._category,
+      WebInspector.resourceTypes.XHR._category
+    ];
+    if (nonCriticalResourceTypes.includes(resourceTypeCategory) ||
+        request.mimeType && request.mimeType.startsWith('image/')) {
+      return false;
+    }
+
+    return ['VeryHigh', 'High', 'Medium'].includes(request.priority());
+  }
+
+  static flattenRequest(record) {
+    const ancestor = record.initiatorRequest();
+
+    return {
+      id: record._requestId,
+      url: record._url,
+      startTime: record.startTime,
+      endTime: record.endTime,
+      responseReceivedTime: record.responseReceivedTime,
+      transferSize: record.transferSize,
+      parent: ancestor ? this.flattenRequest(ancestor) : null,
+    };
+  }
+
+  static extractRequests(networkRecords) {
```

Comment:
This split seems weird to me, shouldn't this belong with the `*Chains` computed artifact and this class just filters to the requests? It's been a while, but I thought we decided to split based on the filtering/chain creation.

The world has also changed quite a bit by now with fast mode introducing a new full-blown dependency graph, instead of all this work we could probably just get away with using the `isCritical` static method on the critical request chains class

Reply:
I still need to drop requests that start from a low priority request. Or am I wrong?

Reply:
if I look at #931 it looks like we just want all critical resources so maybe it makes sense to just look at the priority and not an image and do more logic in the critical-request-chains gatherer

Reply:
Oh right right. Yeah this is good then for now. We'll probably end up migrating this to use the dependency graph from #2720 at some point.

Reply:
OK @wardpeet talked it over with @paulirish and @brendankenny, and we think we should stick to measuring the TTFB for **just the primary HTML request** to keep it simple and avoid messing with the critical request chains gatherers for now. 

We're about to get better infrastructure for dealing with the graph of requests that should make this *much* easier and we can revisit it then. So sorry for the churn here, and my bad for not checking we really needed all the critical requests before suggesting the split 😅  You're a champ 💪

Reply:
yes, you will still need to construct the chain but critical resources would give a flat request tree so it's easier to model other audits/gatherers around it.

 I will revert the changes and only measure ttfb on primary request